{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beef31a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display\n",
    "import os\n",
    "import re\n",
    "import Bio\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffbf714a",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "this will be a notebook on dataset processing only. so merging the different ipr sets\n",
    "then doing basic filtering such as removing sequences without the CxC motif\n",
    "the steps for that are as follows:\n",
    "rename the IPR column from IPR to IPR003572 IPR009908 and IPR012932\n",
    "set the Accession number as the dataframe index\n",
    "merge all three dataframes by using the \"join\" outer function\n",
    "following these steps will combine the data frames in such a way that there will be no duplicate accession numbers\n",
    "and the IPRs will have their own columns, with NAN in places were there is no instance\n",
    "after that the seq strings will be searched for motifs and total cystiene count\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96dcb9f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### dsbb SECTION"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ffd6b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "dsbb_pro_df1 = pd.read_csv(\"/PATH/IPR003752/bacteria_dsbb_set.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84dadfbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(dsbb_pro_df1)\n",
    "dsbb_pro_df2 = dsbb_pro_df1.rename(columns={\"IPR\":\"IPR003752\"}, errors=\"raise\")\n",
    "print(dsbb_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3ca28e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### maue section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "489d3781",
   "metadata": {},
   "outputs": [],
   "source": [
    "maue_pro_df1 = pd.read_csv(\"/PATH/IPR009908/bacteria_maue_set.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9559f703",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(maue_pro_df1)\n",
    "maue_pro_df2 = maue_pro_df1.rename(columns={\"IPR\":\"IPR009908\"}, errors=\"raise\")\n",
    "print(maue_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ae8f033",
   "metadata": {},
   "outputs": [],
   "source": [
    "### vkor section "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfdf1ccf",
   "metadata": {},
   "outputs": [],
   "source": [
    "vkor_pro_df1 = pd.read_csv(\"/PATH/IPR012932/bacteria_vkor_set.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f62988f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(vkor_pro_df1)\n",
    "vkor_pro_df2 = vkor_pro_df1.rename(columns={\"IPR\":\"IPR012932\"}, errors=\"raise\")\n",
    "print(vkor_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e6e38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ribosome section\n",
    "ribo_pro_df1 = pd.read_csv(r\"C:\\PATH\\bacteria_RIBOSOME_set.csv\") #I was unfortunately using a windows PC so this is the norm going forward\n",
    "print(ribo_pro_df1)\n",
    "ribo_pro_df2 = ribo_pro_df1.rename(columns={\"IPR\":\"IPR047873\"}, errors=\"raise\")\n",
    "print(ribo_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25c74782",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rna sub section\n",
    "rna_pro_df1 = pd.read_csv(r\"C:\\PATH\\IPR010243\\bacteria_RNASUB_set.csv\")\n",
    "print(rna_pro_df1)\n",
    "rna_pro_df2 = rna_pro_df1.rename(columns={\"IPR\":\"IPR010243\"}, errors=\"raise\")\n",
    "print(rna_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9909466b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#peptidase section\n",
    "pep_pro_df1 = pd.read_csv(r\"C:\\PATH\\IPR036286\\bacteria_PEPTIDASE_set.csv\")\n",
    "print(pep_pro_df1)\n",
    "pep_pro_df2 = pep_pro_df1.rename(columns={\"IPR\":\"IPR036286\"}, errors=\"raise\")\n",
    "print(pep_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49361e89",
   "metadata": {},
   "outputs": [],
   "source": [
    "#MauG section\n",
    "maug_pro_df1 = pd.read_csv(r\"C:\\PATH\\IPR026259\\bacteria_MAUG_set.csv\")\n",
    "print(maug_pro_df1)\n",
    "maug_pro_df2 = maug_pro_df1.rename(columns={\"IPR\":\"IPR026259\"}, errors=\"raise\")\n",
    "print(maug_pro_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0eb77a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# below is re indexing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95dd2ed2",
   "metadata": {},
   "outputs": [],
   "source": [
    "maue_pro_df3 = maue_pro_df2.set_index('Source_Org_ID')\n",
    "maug_pro_df3 = maug_pro_df2.set_index('Source_Org_ID')\n",
    "dsbb_pro_df3 = dsbb_pro_df2.set_index('Source_Org_ID')\n",
    "vkor_pro_df3 = vkor_pro_df2.set_index('Source_Org_ID')\n",
    "ribo_pro_df3 = ribo_pro_df2.set_index('Source_Org_ID')\n",
    "rna_pro_df3 = rna_pro_df2.set_index('Source_Org_ID')\n",
    "pep_pro_df3 = pep_pro_df2.set_index('Source_Org_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa08a204",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(maue_pro_df3)\n",
    "print(maug_pro_df3)\n",
    "print(dsbb_pro_df3)\n",
    "print(vkor_pro_df3)\n",
    "print(ribo_pro_df3)\n",
    "print(rna_pro_df3)\n",
    "print(pep_pro_df3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2114131a",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_set_df1 = pd.concat([maue_pro_df3,dsbb_pro_df3,vkor_pro_df3,ribo_pro_df3,rna_pro_df3,pep_pro_df3,maug_pro_df3])\n",
    "print(combined_set_df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f7f0cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(combined_set_df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5edc3043",
   "metadata": {},
   "outputs": [],
   "source": [
    "### exporting full data set at this point move to NOTEBOOK 3\n",
    "combined_set_df1.to_excel(r'C:\\PATH\\1_full_set_raw_10_2024.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db5dabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "At this point the notebook breaks off and you need to perform the cxc annotation in order to use the remainder.\n",
    "Onces thats done then you can return to this to merge the data into species. The filter requirements for MauE are only that\n",
    "it has the CXC in the IPR region, and that it has IPR009908. This drops some false positives.\n",
    "\n",
    "ALSO --\n",
    "MauA was added later in the research project because it was found to be a better marker for the MAU cluster then MauG so\n",
    "this is how I added MauA. Basically I am merging it to the set called 2_FBDS_annotated_CXC.xlsx because that wont change any \n",
    "information at this point in the pipeline.\n",
    "\n",
    "and I am adding on to that here in 2_4\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af02a138",
   "metadata": {},
   "outputs": [],
   "source": [
    "maua_pro_df1 = pd.read_csv(r\"C:\\PATH\\IPR036560\\bacteria_MAUA_set.csv\")\n",
    "\n",
    "maua_pro_df2 = maua_pro_df1.rename(columns={\"IPR\":\"IPR036560\"}, errors=\"raise\")\n",
    "\n",
    "\n",
    "fbds = pd.read_excel(r'C:\\PATH\\2_FBDS_annotated_CXC.xlsx')\n",
    "print(fbds.shape)\n",
    "\n",
    "\n",
    "combined_set_df1 = pd.concat([maua_pro_df2,fbds],ignore_index=True, sort=False)\n",
    "print(combined_set_df1.shape)\n",
    "print(fbds.columns)\n",
    "print(combined_set_df1.columns)\n",
    "\n",
    "#this is the resulting set that moved on \n",
    "combined_set_df1.to_excel(r'C:\\PATH\\3_FBDS_annotated_CXC_MAUA.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c26a8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#I saved a copy here and reloaded it\n",
    "combined_set_df1 = pd.read_excel(r'C:\\PATH\\3_FBDS_annotated_CXC_MAUA.xlsx')\n",
    "\n",
    "cols = list(combined_set_df1.columns.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73017326",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0164d87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dropping some columns\n",
    "combined_set_df2=combined_set_df1[[ 'Accession', 'name', 'gene', 'protein_length', 'in_alphafold', 'Source_Org_Name', 'taxa_range', 'Sequence', 'superkingdom', 'clade', 'phylum', 'class', 'order', 'family', 'genus', 'species', 'strain', 'no rank', 'species group', 'subspecies', 'suborder', 'subclass', 'subgenus', 'species subgroup', 'isolate', 'tribe', 'pathogroup', 'subfamily', 'serotype', 'serogroup', 'forma specialis', 'biotype','CXC_at_all', 'CXC_past', 'CXC_before', 'CXC_inside_IPR', 'IPR012932',  'IPR003752', 'IPR009908','IPR047873', 'IPR010243', 'IPR036286','IPR026259','IPR036560']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14fb68d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "display(combined_set_df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a0ff04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#removing MauE entries that do not satisfy the CXC motif being inside the IPR region by splitting and filtering\n",
    "maue_cxc_mask = (combined_set_df2['IPR009908'] == 'IPR009908') & (combined_set_df2['CXC_inside_IPR'] > 0)\n",
    "true_maue = combined_set_df2.loc[maue_cxc_mask]\n",
    "rest_of_set_mask = (combined_set_df2['IPR009908'] != 'IPR009908')\n",
    "rest_of_set = combined_set_df2.loc[rest_of_set_mask]\n",
    "display(true_maue)\n",
    "\n",
    "display(rest_of_set)\n",
    "                                                                \n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d77a7de",
   "metadata": {},
   "outputs": [],
   "source": [
    "#merging into the filtered set\n",
    "filtered_fbds = pd.concat([true_maue,rest_of_set])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7c075c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#small check of last row\n",
    "display(filtered_fbds.loc[[314373]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f3fdc5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "The above data set is done. but species without the three quality markers are still present, the below sections give a species \n",
    "level data set that gets filtered for the 3 markers in notebook 5\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f45a6f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "#were condensing down to species counts so were dropping columns not related to phylogeny and IPS counts. at this point\n",
    "#we only have true maue according to our filters\n",
    "phylogeny_columns = ['superkingdom', 'clade', 'phylum', 'class', 'order', 'family', 'genus', 'strain', 'no rank', 'species group', 'subspecies', 'suborder', 'subclass', 'subgenus', 'species subgroup', 'isolate', 'tribe', 'pathogroup', 'subfamily', 'serotype', 'serogroup', 'forma specialis', 'biotype']  # Add other relevant columns here\n",
    "\n",
    "# Group by species and count the IPR entries for that species\n",
    "ipr_counts = filtered_fbds.groupby('species')[['IPR009908', 'IPR012932', 'IPR003752','IPR047873', 'IPR010243', 'IPR036286','IPR026259','IPR036560']].count()\n",
    "display(ipr_counts)\n",
    "# pulls the phylogeny info by taking the first Taxid for each phylo column for the grouped species. they are all the same.\n",
    "phylogeny_info = combined_set_df2.groupby('species')[phylogeny_columns].first()\n",
    "display(phylogeny_info)\n",
    "# Combine data frames\n",
    "species_summary_df = pd.concat([ipr_counts, phylogeny_info], axis=1)\n",
    "\n",
    "# reset the index so we keep species as a column\n",
    "species_summary_df.reset_index(inplace=True)\n",
    "\n",
    "# small check\n",
    "print(species_summary_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "998eca7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#export the data\n",
    "filtered_fbds.to_excel(r\"C:\\PATH\\3b_FBDS_CXC_annotated_false_pos_removed_MAUA.xlsx\")\n",
    "species_summary_df.to_excel(r\"C:\\PATH\\4_Filtered_FBDS_IPRcounts_byspecies_MAUA.xlsx\")\n",
    "#youll have to fill na to zero be careful if you using the .count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
