{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab4d42fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "I have to go back and remake the logo plot to take into account every single MauE with a CXC within the IPR region\n",
    "To make this a bit easier I am going to limit this to those with a single CXC since there are about 300 ish 2-3 and a hand full\n",
    "with more\n",
    "\n",
    "I will essentially go by the species list of species with the 3 quality markers. use that to filter down the list of MauE\n",
    "with 1 CXC so its only things with the 3 quality markers used in the phylogenetic anaysis \n",
    "\n",
    "then ill use regex to just pull the CXC and two amino acids on either side. the dataframe will be the accession number \n",
    "then the then I will create a Fasta file that is just the same information so i can work it into the rest of my code\n",
    "basically the final aligned file will be\n",
    "\n",
    "the logo plot\n",
    ">asdf\n",
    "XXCXCXX\n",
    ">asdf\n",
    "XXCXCXX\n",
    ".\n",
    ".\n",
    ".\n",
    "\n",
    "this will then be made into a logo plot using logomaker. If the CXC is at a terminus and therefor wont have a position before \n",
    "or after the CXC it needs to be dropped. This turned out to only be 5 MauE in the whole set.\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55084a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import re\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f81a2283",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this was a species list from the species level set that only has species with the three quality markers present\n",
    "#then drops mauE entries from species not on the list\n",
    "species = pd.read_excel(r\"C:\\PATH\\5_Filtered_FBDS_IPRcounts_byspecies_MAUA_names.xlsx\")\n",
    "fbds = pd.read_excel(r\"C:\\PATH\\3b_FBDS_CXC_annotated_false_pos_removed_MAUA.xlsx\")\n",
    "print(species.shape)\n",
    "print(fbds.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96348a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "#i am only looking at 1 cxc MauE\n",
    "maue = fbds[fbds['CXC_at_all'] == 1].copy()\n",
    "print(maue.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f11cbe5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = list(species['species'])\n",
    "print(species_list[0:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f004e26e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#in this section I am dropping anyting not in the species list I just made\n",
    "#I also need to remove things that have the CXC at the N or C terminus because they wont be able to actually align \n",
    "#becuase I am including two positions before and after the CXC\n",
    "species_count = 0\n",
    "with open(r'C:\\PATH\\XXCXCXX_9925.fasta','a') as file:\n",
    "    for i in maue.index:\n",
    "        spec = maue.at[i,'species']\n",
    "        accession = maue.at[i,'Accession']\n",
    "        sequence = maue.at[i,'Sequence']\n",
    "        search = re.search(r\"[A-ZU][A-ZU]C[A-ZU]C[A-ZU][A-ZU]\",sequence)\n",
    "        if search:\n",
    "            cxc = search.group()\n",
    "            if spec in species_list:\n",
    "                file.write('>'+accession+'\\n'+cxc+'\\n')\n",
    "                species_count += 1\n",
    "            else:\n",
    "                continue\n",
    "        else:\n",
    "            print(accession,' has a CXC at either end')\n",
    "    file.close()\n",
    "print(species_count)\n",
    "#delete the file as it will just append to any file with the same name if your repeating        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda89441",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "the results of the above code at the time of writing listed these five MauE:\n",
    "A0A357LZM8  has a CXC at either end\n",
    "A0A3D5AI03  has a CXC at either end\n",
    "A0A6G3AJL1  has a CXC at either end\n",
    "A0A838DJ16  has a CXC at either end\n",
    "A0A970UZD1  has a CXC at either end\n",
    "\n",
    "which have a CXC at the C or N terminal end and thus dont have flanking amino acids. they still seems to place\n",
    "the motif in the space between the membranes in their predicted fold but i cant use them for a logo plot\n",
    "\n",
    "this fasta file can now be used directly to make the logo plot using logo maker package for python and biopythin ALignIO\n",
    "'''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
